rhp/v3/rhp.go

```diff
@@ -87,134 +96,145 @@ func PayByContract(rev *types.FileContractRevision, amount types.Currency, refun
-
-	// Cost values specific to the RenewContract instruction.
-	RenewContractCost types.Currency `json:"renewcontractcost"`
-
-	// Cost values specific to the Revision command.
-	RevisionBaseCost types.Currency `json:"revisionbasecost"`
-
-	// SwapSectorCost is the cost of swapping 2 full sectors by root.
-	SwapSectorCost types.Currency `json:"swapsectorcost"`
-
-	// Cost values specific to the Write instruction.
-	WriteBaseCost   types.Currency `json:"writebasecost"`   // per write
-	WriteLengthCost types.Currency `json:"writelengthcost"` // per byte written
-	WriteStoreCost  types.Currency `json:"writestorecost"`  // per byte / block of additional storage
-
-	// TxnFee estimations.
-	TxnFeeMinRecommended types.Currency `json:"txnfeeminrecommended"`
-	TxnFeeMaxRecommended types.Currency `json:"txnfeemaxrecommended"`
-
-	// ContractPrice is the additional fee a host charges when forming/renewing
-	// a contract to cover the miner fees when submitting the contract and
-	// revision to the blockchain.
-	ContractPrice types.Currency `json:"contractprice"`
-
-	// CollateralCost is the amount of money per byte the host is promising to
-	// lock away as collateral when adding new data to a contract. It's paid out
-	// to the host regardless of the outcome of the storage proof.
-	CollateralCost types.Currency `json:"collateralcost"`
-
-	// MaxCollateral is the maximum amount of collateral the host is willing to
-	// put into a single file contract.
-	MaxCollateral types.Currency `json:"maxcollateral"`
-
-	// MaxDuration is the max duration for which the host is willing to form a
-	// contract.
-	MaxDuration uint64 `json:"maxduration"`
-
-	// WindowSize is the minimum time in blocks the host requests the
-	// renewWindow of a new contract to be.
-	WindowSize uint64 `json:"windowsize"`
-
-	// Registry related fields.
-	RegistryEntriesLeft  uint64 `json:"registryentriesleft"`
-	RegistryEntriesTotal uint64 `json:"registryentriestotal"`
-}
+type (
+	// A SettingsID is a unique identifier for registered host settings used by renters
+	// when interacting with the host.
+	SettingsID [16]byte
+
+	// ResourceCost is the cost of executing an instruction.
+	ResourceCost struct {
+		// Base is the cost to execute the instruction. It is not refunded if
+		// the instruction fails.
+		Base types.Currency
+		// Storage is the cost to store the instruction's data. Storage is the
+		// additional storage usage in bytes used. It is refunded if the
+		// program fails.
+		Storage types.Currency
+		// Collateral is the amount of collateral that must be risked by the
+		// host to execute the instruction. It is freed if the instruction
+		// fails.
+		Collateral types.Currency
+		// Egress is the cost to send the instruction's output to the renter. It
+		// is not refunded if the program fails.
+		Egress types.Currency
+		// Ingress is the cost to receive the instruction's input from the
+		// renter. It is not refunded if the program fails.
+		Ingress types.Currency
+	}
```

Comment:
how about:

```go
ResourceUsage struct {
    Base types.Currency
    Collateral types.Currency
    Storage uint64
    Ingress uint64
    Egress uint64
}
```
Then, methods like `AppendSectorCost` can become standalone `AppendSectorUsage` functions, and we can have a single `func(pt HostPriceTable, ru ResourceUsage) types.Currency` that multiplies each field by its corresponding cost.

Regardless, I imagine we'll want an `Add` method for adding usages/costs together, right? For programs with multiple instructions?

Reply:
I don't have a strong opinion either way, but you'll still have outliers like `ReadOffsetCost` and `ReadSectorCost` which require `length`